File: backend/app/services/static_map.py
"""Google Static Maps URL builder (uses GOOGLE_MAPS_API_KEY from backend .env)."""

import os
from typing import Any

MAP_W = 640
MAP_H = 400


def _bounds(
    markers: list[dict[str, Any]],
    user_lat: float | None,
    user_lng: float | None,
) -> dict[str, float]:
    lats = [float(m["lat"]) for m in markers]
    lngs = [float(m["lng"]) for m in markers]
    if user_lat is not None:
        lats.append(user_lat)
    if user_lng is not None:
        lngs.append(user_lng)
    pad = 0.008
    return {
        "min_lat": min(lats) - pad,
        "max_lat": max(lats) + pad,
        "min_lng": min(lngs) - pad,
        "max_lng": max(lngs) - pad,
        "center_lat": (min(lats) + max(lats)) / 2,
        "center_lng": (min(lngs) + max(lngs)) / 2,
    }


def build_google_static_map_url(
    markers: list[dict[str, Any]],
    user_lat: float | None = None,
    user_lng: float | None = None,
) -> str | None:
    api_key = os.getenv("GOOGLE_MAPS_API_KEY") or os.getenv("GOOGLE_API_KEY")
    if not api_key or not markers:
        return None

    b = _bounds(markers, user_lat, user_lng)
    parts = [
        f"center={b['center_lat']},{b['center_lng']}",
        "zoom=14",
        f"size={MAP_W}x{MAP_H}",
        "scale=2",
        "maptype=roadmap",
    ]
    if user_lat is not None and user_lng is not None:
        parts.append(f"markers=color:green%7Clabel:Y%7C{user_lat},{user_lng}")
    for i, m in enumerate(markers):
        color = "orange" if m.get("is_recommended") else "violet"
        label = str((i % 9) + 1)
        parts.append(
            f"markers=color:{color}%7Clabel:{label}%7C{m['lat']},{m['lng']}"
        )
    return f"https://maps.googleapis.com/maps/api/staticmap?{'&'.join(parts)}&key={api_key}"

File: khidmat-ai/mobile/lib/staticMap.ts
import axios from 'axios';
import type { MapMarker } from '../api/client';

import { getApiBaseUrl } from './apiConfig';

const MAP_W = 640;
const MAP_H = 400;

function bounds(markers: MapMarker[], userLat?: number, userLng?: number) {
  const lats = markers.map((m) => m.lat);
  const lngs = markers.map((m) => m.lng);
  if (userLat != null) lats.push(userLat);
  if (userLng != null) lngs.push(userLng);
  const pad = 0.008;
  return {
    minLat: Math.min(...lats) - pad,
    maxLat: Math.max(...lats) + pad,
    minLng: Math.min(...lngs) - pad,
    maxLng: Math.max(...lngs) + pad,
    centerLat: (Math.min(...lats) + Math.max(...lats)) / 2,
    centerLng: (Math.min(...lngs) + Math.max(...lngs)) / 2,
  };
}

/** Google Static Maps — set EXPO_PUBLIC_GOOGLE_MAPS_API_KEY in mobile/.env */
export function googleStaticMapUrl(
  markers: MapMarker[],
  userLat?: number,
  userLng?: number
): string | null {
  const key =
    process.env.EXPO_PUBLIC_GOOGLE_MAPS_API_KEY ||
    process.env.EXPO_PUBLIC_GOOGLE_API_KEY;
  if (!key || !markers.length) return null;

  const b = bounds(markers, userLat, userLng);
  const parts: string[] = [
    `center=${b.centerLat},${b.centerLng}`,
    'zoom=14',
    `size=${MAP_W}x${MAP_H}`,
    'scale=2',
    'maptype=roadmap',
  ];

  if (userLat != null && userLng != null) {
    parts.push(`markers=color:green%7Clabel:Y%7C${userLat},${userLng}`);
  }
  markers.forEach((m, i) => {
    const color = m.is_recommended ? 'orange' : 'violet';
    const label = String((i % 9) + 1);
    parts.push(`markers=color:${color}%7Clabel:${label}%7C${m.lat},${m.lng}`);
  });

  return `https://maps.googleapis.com/maps/api/staticmap?${parts.join('&')}&key=${key}`;
}

/** Uses GOOGLE_MAPS_API_KEY from backend/.env (no mobile key required). */
export async function fetchGoogleStaticMapFromApi(
  markers: MapMarker[],
  userLat?: number,
  userLng?: number
): Promise<string | null> {
  if (!markers.length) return null;
  try {
    const { data } = await axios.post<{ url: string }>(
      `${getApiBaseUrl()}/api/maps/static`,
      {
        markers,
        user_lat: userLat,
        user_lng: userLng,
      },
      { timeout: 12000 }
    );
    return data.url || null;
  } catch {
    return null;
  }
}

/** Client env key first, then backend API. */
export async function resolveGoogleMapUrl(
  markers: MapMarker[],
  userLat?: number,
  userLng?: number
): Promise<string | null> {
  const local = googleStaticMapUrl(markers, userLat, userLng);
  if (local) return local;
  return fetchGoogleStaticMapFromApi(markers, userLat, userLng);
}

/** Free embed — no API key (web iframe). */
export function openStreetMapEmbedUrl(
  markers: MapMarker[],
  userLat?: number,
  userLng?: number
): string | null {
  if (!markers.length) return null;
  const b = bounds(markers, userLat, userLng);
  const lat = userLat ?? b.centerLat;
  const lng = userLng ?? b.centerLng;
  return (
    `https://www.openstreetmap.org/export/embed.html` +
    `?bbox=${b.minLng}%2C${b.minLat}%2C${b.maxLng}%2C${b.maxLat}` +
    `&layer=mapnik&marker=${lat}%2C${lng}`
  );
}
